### PR TITLE
fix: Prevent newExporter crash if tree.ndb is nil

### PR DESCRIPTION
**Pull request by chillyvee**
Found a case where dig chain did not have all stores configured with a nodedb (for whatever reason)

It is better to return an error so that cosmos-sdk can determine whether to skip the misconfigured store or panic.

Would also like to backport to fix v0.19.4 (but need assistance to target the right branch)

### CODE REVIEW

File: export.go [hunk 1]
```diff
@@ -34,6 +35,15 @@ type Exporter struct {
 
 // NewExporter creates a new Exporter. Callers must call Close() when done.
 func newExporter(tree *ImmutableTree) *Exporter {
+	if tree == nil {
+		return nil
+	}
+	// CV Prevent crash on incrVersionReaders if tree.ndb == nil
+	if tree.ndb == nil {
+		fmt.Printf("iavl/export newExporter failed to create because tree.ndb is nil\n")
```
Comment (robert-zaremba) [inline export.go:43]: this function should return error. Doing `fmt.Printf` doesn't make sense here.

Comment (tac0turtle) [reply export.go:43]: agree, lets break the api.

Comment (chillyvee) [reply export.go:43]: I can just return (nil, err) with fmt.Errorf to start with.  Then any more complicated error handling can be followed up in other commits/improvment?

Or is there some error standard you're trying to maintain?

Comment (tac0turtle) [reply export.go:43]: yea that makes sense, we would then handle this in the application

Comment (chillyvee) [reply export.go]: Returning error now, updated api/test

**Comment by chillyvee** [comment]
Last commit returns explicit error as recommended.
Tests also updated

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 ## Unreleased
 
+- [#622](https://github.com/cosmos/iavl/pull/622) `export/newExporter()` and `ImmutableTree.Export()` returns error for nil arguements
```
Comment (robert-zaremba) [inline CHANGELOG.md:5]: this should be put in `### Breaking Chagnes`

File: export.go [hunk 3]
```diff
@@ -33,7 +39,15 @@ type Exporter struct {
...
 
 // NewExporter creates a new Exporter. Callers must call Close() when done.
-func newExporter(tree *ImmutableTree) *Exporter {
+func newExporter(tree *ImmutableTree) (*Exporter, error) {
+	if tree == nil {
+		return nil, ErrorTreeNil
+	}
+	// CV Prevent crash on incrVersionReaders if tree.ndb == nil
+	if tree.ndb == nil {
+		return nil, ErrorTreeNodeDbNil
```
Comment (robert-zaremba) [inline export.go:48]: maybe we can have one error type: `ErrNotIniitalizedTree`. And then use wrapping to add more details, eg:
```
ErrNotIniitalizedTree.Wrap("ndb is nil")
```

**Review by robert-zaremba — COMMENTED** [review]
lgtm, added one suggestion

File: CHANGELOG.md (hunk 2 above)
Comment (chillyvee) [reply CHANGELOG.md:5]: Moved

**Comment by chillyvee** [comment]
change log updated, errors wrapped

**Comment by chillyvee** [comment]
note change in errors package

File: export.go [hunk 4]
```diff
@@ -33,7 +37,15 @@ type Exporter struct {
 }
 
 // NewExporter creates a new Exporter. Callers must call Close() when done.
-func newExporter(tree *ImmutableTree) *Exporter {
+func newExporter(tree *ImmutableTree) (*Exporter, error) {
+	if tree == nil {
+		return nil, errors.Wrap(ErrNotInitalizedTree, "tree is nil")
```
Comment (tac0turtle) [inline export.go:42]: can we use fmt.Errorf to avoid introducing GitHub.com/pkg/errors as a dependency

Comment (chillyvee) [reply export.go:42]: My mistake.  I thought the team was looking to call Wrap() specifically.  Will change it again.

**Comment by chillyvee** [comment]
Updated : [change from requested Wrap to fmt.Errorf](https://github.com/cosmos/iavl/pull/622/commits/e856bd2f65050eeac52c39d4ce080f586c37bd4f)

File: export.go (hunk 4 above)
Comment (tac0turtle) [reply export.go]: no worries, pkg/errors is deprecated with go 1.18 and beyond because of fmt.errorf and/or errors pkg.

**Review by tac0turtle — APPROVED** [review]
thank you for submitting this!!! we can look at back porting this to avoid the issues being seen.

**Comment by tac0turtle** [comment]
linting and tests are failing, could you fix and we will handle backport

**Comment by chillyvee** [comment]
> linting and tests are failing, could you fix and we will handle backport

Will give it a shot and let you know if help is required :)

File: export.go (hunk 4 above)
Comment (chillyvee) [reply export.go]: Learn something new every day

**Comment by chillyvee** [comment]
Issue was hiding in the benchmarks, fixed and running through tests.

**Comment by chillyvee** [comment]
looks like there's some fumpting to do

**Comment by chillyvee** [comment]
Also as a note, noticed:
```
dfde9a5 iavl: remove grpc server (#499) [GitHub]
0       129     cmd/iavlserver/README.md
0       207     cmd/iavlserver/main.go
```

But in Makefile it still tries to build iavlserver
```
ifeq ($(COLORS_ON),)
        go install ./cmd/iaviewer
        go install ./cmd/iavlserver <-- FAILS
else
        go install $(CMDFLAGS) ./cmd/iaviewer
        go install $(CMDFLAGS) ./cmd/iavlserver <-- FAILS
endif
```

Want it removed?

**Comment by chillyvee** [comment]
Removed failing build of cmd/iavlserver.  Let me know if you want it back.

**Comment by tac0turtle** [comment]
thanks for the pr @chillyvee

**Comment by tac0turtle** [comment]
@mergifyio backport release/v0.19.x

**Comment by tac0turtle** [comment]
@Mergifyio backport release/v0.19.x

**Pull request merged**